sycl/test-e2e/ESIMD/unified_memory_api/Inputs/atomic_update_slm.hpp

```diff
@@ -0,0 +1,720 @@
+            << " n_groups=" << n_groups << " start_ind=" << start_ind
+            << " masked_lane=" << masked_lane << " repeat=" << repeat
+            << " stride=" << stride << " }...";
+
+  constexpr size_t size = start_ind + (N - 1) * stride + 1;
+  T *arr = malloc_shared<T>(size, q);
+  constexpr int n_threads = threads_per_group * n_groups;
+
+  for (int i = 0; i < size; ++i) {
+    arr[i] = ImplF<T, N>::init(i);
+  }
+
+  range<1> glob_rng(n_threads);
+  range<1> loc_rng(threads_per_group);
+  nd_range<1> rng(glob_rng, loc_rng);
+
+  try {
+    auto e = q.submit([&](handler &cgh) {
+      cgh.parallel_for(rng, [=](id<1> ii) SYCL_ESIMD_KERNEL {
+        int i = ii;
+        slm_init<32768>();
+        simd<uint32_t, N> offsets(start_ind * sizeof(T), stride * sizeof(T));
+        simd<T, size> data;
+        data.copy_from(arr);
+
+        simd<uint32_t, size> slm_offsets(0, sizeof(T));
+        // TODO: replace this API with unified version.
+        lsc_slm_scatter(slm_offsets, data);
+
+        simd_mask<N> m = 1;
+        if constexpr (UseMask) {
+          if (masked_lane < N)
+            m[masked_lane] = 0;
+        }
+        // Intra-work group barrier.
+        barrier();
+
+        // the atomic operation itself applied in a loop:
+        for (int cnt = 0; cnt < repeat; ++cnt) {
+          if constexpr (n_args == 0) {
+            if constexpr (UseMask) {
+              slm_atomic_update<op, T>(offsets, m);
+            } else {
+              slm_atomic_update<op, T>(offsets);
+            }
+          } else if constexpr (n_args == 1) {
+            simd<T, N> v0 = ImplF<T, N>::arg0(i);
+            if constexpr (UseMask) {
+              slm_atomic_update<op, T>(offsets, v0, m);
+            } else {
+              slm_atomic_update<op, T>(offsets, v0);
+            }
+          } else if constexpr (n_args == 2) {
+            simd<T, N> new_val = ImplF<T, N>::arg0(i); // new value
+            simd<T, N> exp_val = ImplF<T, N>::arg1(i); // expected value
+            // do compare-and-swap in a loop until we get expected value;
+            // arg0 and arg1 must provide values which guarantee the loop
+            // is not endless:
+            if constexpr (UseMask) {
+              for (simd<T, N> old_val =
+                       slm_atomic_update<op, T>(offsets, new_val, exp_val, m);
+                   any(old_val < exp_val, !m);
+                   old_val =
+                       slm_atomic_update<op, T>(offsets, new_val, exp_val, m))
+                ;
+            } else {
+              for (simd<T, N> old_val =
+                       slm_atomic_update<op, T>(offsets, new_val, exp_val);
+                   any(old_val < exp_val, !m);
+                   old_val =
+                       slm_atomic_update<op, T>(offsets, new_val, exp_val))
+                ;
+            }
+          }
+        }
+        // TODO: replace this API with unified version.
+        auto data0 = lsc_slm_gather<T>(slm_offsets);
```

Comment:
same as for lsc_slm_scatter above.
You can use slm_block_load() here instead of lsc_slm_gather

Reply:
Done.